issue: 2597798 Add observation based measurement, more stats, and histogram 2

ChrisCoe opened the pull request:
**Motive:**

Currently, sockperf runs until a timer completes which produces an undetermined number of data rows. Observational-based measurements produce an exact number which eases sockperf run comparisons (such as with stats) and avoids potential overflow during high data collection. Compute more stats directly from measurements to take advantage of sockperf's `ticks` (nanosecond precision). Produce microsecond resolution sparse fixed bin histogram from measurements to provide a visual display of data.

**Changes:**
Obervational-based
- Parser accepts `-n` for number of observations. This measurement type is only for ping-pong mode
- Observational warmup/cooldown when in observational mode instead of adding warmup/cooldown time to timer
- Tracking valid observations in `doSendThenReceiveLoop`, where `client_send_then_receive` is blocking
- Time cap based on warm up sample to limit waiting on observations

Stats
- Add mean absolute deviation, median absolute deviation, semi-interquartile range, coefficient of variance, standard error, and confidence interval 
- Confidence interval significance level configurable via arg flag `--ci_sig_level` (includes decimals)
- Inverse CDF (or ppf) function to compute confidence interval  for any significance level (0-100 exclusive)
- Getter for long name of options for logging purposes
- Sort latency list once where other stats can reuse

Histogram
- Build histogram and display histogram on terminal for a quick view (frequency rounded up)
- Store raw histogram data to log file if present
- Histogram flags for histogram bin size, lower range, and upper range of interest configurable via cmd respectively `-histogram {bin size}:{lower range}:{upper range}`
- Left and right outlier bins to capture all data outside given range

**Manual Self-Testing:**
Ran server and client locally and manually verified logX.txt had the expected observational data count.
`./sockperf sr -i 127.0.0.1`
`./sockperf pp -i 127.0.0.1 -n 1000 -m 64 --full-log log.txt` -> 1000
`./sockperf pp -i 127.0.0.1 -n 1         -m 64 --full-log logA.txt` -> 1\
`./sockperf pp -i 127.0.0.1 -n -1        -m 64 --full-log logC.txt` -> failed to help page as expected
`./sockperf pp -i 127.0.0.1 -n 100000001 -m 64 --full-log logD.txt` -> failed to help page as expected
`./sockperf pp -i 127.0.0.1 -n 100000000 -m 64 --full-log logE.txt` -> 100000000 (max obs imposed)


`./sockperf pp -i 127.0.0.1 -n 1000 -m 64 --ci_sig_level 80 --full-log log.txt --full-rtt`
All stat changes are reported in magenta line `sockperf: ====> avg-rtt...`.
![image](https://user-images.githubusercontent.com/20761371/108804256-c3639a80-756a-11eb-963c-53ea3cbccb45.png)


`./sockperf pp -i 127.0.0.1 -n 1000000 -m 64 --histogram 1:25:60 --full-log log.txt --full-rtt`

All histogram changes are after `sockperf: [Histogram]...`
![image](https://user-images.githubusercontent.com/20761371/133393025-78f18cfd-767a-448b-a949-300a0dd6cf49.png)

Raw histogram content at the end of log file:
```
------------------------------
histogram was built using the following parameters: --h_bin_size_us=1 --h_lower_range_us=25 --h_upper_range_us=60
------------------------------
bin (usec)            frequency
36                         1
37                         1
38                         343
39                         2555
40                         1715
41                         1712
42                         2635
43                         25658
44                         41436
45                         181808
46                         84039
47                         31985
48                         27579
49                         127625
50                         180704
51                         61690
52                         23414
53                         14486
54                         11610
55                         10912
56                         9841
57                         7917
58                         6856
59                         5880
60-3938                    137598
------------------------------
```

ChrisCoe committed:
Add obervation-based measurement for ping pong

ChrisCoe committed:
Add more stats

ChrisCoe committed:
Add sparce fixed bin histogram

swx-jenkins2 commented:
Can one of the admins verify this patch?

ChrisCoe committed:
Handle histogram display when no terminal size

ChrisCoe commented:
Hi @igor-ivanov, did you get a chance to review this PR?

igor-ivanov commented:
@ChrisCoe I will try to do it this week

igor-ivanov commented on src/sockperf.cpp:
```diff
@@ -725,6 +877,24 @@ static int proc_mode_ping_pong(int id, int argc, const char **argv) {
 
     /* Set command line specific values */
     if (!rc && self_obj) {
+        if (!rc && aopt_check(self_obj, 'n')) {
```
As I understand `-n` conflicts with `-t`, so please consider explicit check in both cases. See example as
https://github.com/Mellanox/sockperf/blob/sockperf_v2/src/sockperf.cpp#L1573-L1592.

Folowing test stucks on my side.
```
sockperf sr -i 1.1.3.6
sockperf pp -i 1.1.3.6 --mps=2 -n 10
```

Common note:
1. Idea with `-n` is clear
2. I think that it can conflict with some other parameters as `--mps`, `--time` for example
3. Probably resolving conflicts is complex task
4. Consider replacing **observation** with **number** in all places (functions, variables etc). Number name is more obvious. See other benchmark tools for example iperf3 (https://iperf.fr/iperf-doc.php#3num). Observation confuses.

igor-ivanov commented on src/defs.h:
```diff
@@ -648,6 +665,8 @@ struct user_params_t {
     uint32_t cooldown_msec;
     uint32_t warmup_msec;
+    uint64_t cooldown_obs;
+    uint64_t warmup_obs;
```
using `uint64_t` for warmup and cooldown comparing with `uint32_t` for counter looks strange

igor-ivanov commented on src/sockperf.cpp:
```diff
@@ -138,9 +138,9 @@ static const struct app_modes {
-        "on timeline and message size." },
+      { proc_mode_playback, "playback",
+        aopt_set_string("pb"), "Run " MODULE_NAME " client for latency test using playback of predefined "
+                               "traffic, based on timeline and message size." },
```
potential style check issue

ChrisCoe replied:
I don't think this is an issue. I made this formatting change for better alignment in regards to the rest of the modes in the list

igor-ivanov commented on src/aopt.cpp:
```diff
@@ -297,6 +297,59 @@ const char *aopt_value(const AOPT_OBJECT *aopt_obj, int key) {
+    if (key == 0) {
+        log_err("Not a valid key for aopt");
+        return NULL;
+    }
```
please do check after initialization

igor-ivanov commented on src/client.cpp:
```diff
@@ -64,33 +67,236 @@ void set_client_timer(struct itimerval *timer) {
+    int maxDisplayWidth = 0;
+    std::string prefixToHistogramDisplay ("sockperf: bin XXX-XXX");
+    std::map<int, int>::iterator itr;
+
```
Please fix format output (alignment);
```
sockperf: bin 80-90 #
sockperf: bin 100-110 ####
sockperf: bin 110-120 ######
```

igor-ivanov commented on src/defs.h:
```diff
@@ -222,6 +227,11 @@ enum {
+    OPT_HISTOGRAM,                // 46
+    OPT_HISTOGRAM_UPPER_RANGE,    // 47
+    OPT_HISTOGRAM_LOWER_RANGE,    // 48
+    OPT_HISTOGRAM_BIN_SIZE,       // 49
```
Please consider using `--histogram` option to set all related parameters. See `set_affinity_list` as an example.
I think this manner should look better.
For example: `--histogram bin:lower:upper`

igor-ivanov reviewed (commented):
Please look my comments

ChrisCoe committed:
Combine histogram flags to one

ChrisCoe committed:
Align printed histogram

ChrisCoe committed:
Initialize before check

ChrisCoe committed:
Make obs counter uint64 and rename to target

ChrisCoe committed:
Catch conflicts with -n and -t and --mps options

ChrisCoe committed:
Replace observation name with number

ChrisCoe committed:
minor formatting change

ChrisCoe commented:
@igor-ivanov I addressed all your feedback and agree with all your suggestions; thanks for the provided examples. Meant to do this much sooner... let me know what you think.

igor-ivanov commented:
thank you, @ChrisCoe I will look

ChrisCoe commented:
@igor-ivanov Is there an update on this?

igor-ivanov commented:
@ChrisCoe still no, i will look and back with feedback

igor-ivanov commented:
bot:retest

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -725,19 +831,48 @@ static int proc_mode_ping_pong(int id, int argc, const char **argv) {
                 }
             } else {
-                log_msg("'-%c' Invalid value", 't');
+                log_msg("-t conflicts with -n option");
```
The conflict between `-t` and `-n` has already been checked in previous `if` (lines 834-855).

igor-ivanov replied:
it is true but i would like to keep this too. To avoid a case when someone moves this code before `-n` option parsing.

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -595,6 +606,86 @@ static int proc_mode_under_load(int id, int argc, const char **argv) {
+                    } else if (*cur_ptr == ':') {
+                        *cur_ptr = '\0';
+                        parsed_arg[arg_index] = strtol(cur_buf, &end_ptr, 0);
+                        arg_index++;
```
Possible out-of-bounds problem. For example, try to parse string `12:34:56:78:90`.

I suggest you to rewrite parsing code using sscanf:
```c
long bin_size, lower_range, upper_range;
char suffix; //< needed to check for garbage at the end
if (sscanf(optarg, "%ld:%ld:%ld%c", &bin_size, &lower_range, &upper_range, &suffix) == 3) {
    s_user_params.histogram_bin_size = bin_size;
    s_user_params.histogram_lower_range = lower_range;
    s_user_params.histogram_upper_range = upper_range;
} else {
    log_err("Invalid argument: %s "
            "Format should be binsize:lowerrange:upperrange", optarg);
    rc = SOCKPERF_ERR_BAD_ARGUMENT;
}
```

ChrisCoe replied:
Ah, so much cleaner!

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -3248,6 +3563,11 @@ int bringup(const int *p_daemonize) {
                                   10 * s_user_params.reply_every; // + 10 replies for safety
         _maxSequenceNo += s_user_params.burst_size; // needed for the case burst_size > mps
 
+        if(s_user_params.measurement == NUMBER_BASED) {
```
Please add a space betweeen `if ` and `(`

agalanin-at-nvidia reviewed (changes requested):
Hello, @ChrisCoe! I have some comments regarding parsing code.

agalanin-at-nvidia commented on src/aopt.cpp:
```diff
@@ -297,6 +297,59 @@ const char *aopt_value(const AOPT_OBJECT *aopt_obj, int key) {
+        return NULL;
+    }
+
+    while(aopt_desc[i].key != 0) {
```
It is more common to use `for` loop in this case:
```C
for (int i = 0; aopt_desc[i].key != 0; ++i) {
```

agalanin-at-nvidia commented on src/aopt.cpp:
```diff
@@ -297,6 +297,59 @@ const char *aopt_value(const AOPT_OBJECT *aopt_obj, int key) {
+        return NULL;
+    }
+
+    return *aopt_get_desc(aopt_desc, key)->longs;
```
Check return value of `aopt_get_desc` before dereferencing a pointer. It may be NULL in case of invalid key.

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -64,33 +67,268 @@ void set_client_timer(struct itimerval *timer) {
+int getStartOfRightOutlierBin(){
+    const int lowerRange = s_user_params.histogram_lower_range;
+    const int upperRange = s_user_params.histogram_upper_range;
+    const int binSize = s_user_params.histogram_bin_size;
```
It is better to use the same type (uint32_t) as in `defs.h`. The same remark for `getRightOutlierBinIndexReserved`.

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -64,33 +67,268 @@ void set_client_timer(struct itimerval *timer) {
+}
+
+//------------------------------------------------------------------------------
+int getNumberOfDigits(int input) {
```
uint32_t input

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -64,33 +67,268 @@ void set_client_timer(struct itimerval *timer) {
+            "--h_bin_size_us=%d --h_lower_range_us=%d --h_upper_range_us=%d\n",
+            (int)g_pApp->m_const_params.histogram_bin_size,
+            (int)g_pApp->m_const_params.histogram_lower_range,
+            (int)g_pApp->m_const_params.histogram_upper_range);
```
Replace `%d` with `" PRIu32 "` from `<inttypes.h>` to avoid need to cast types.

ChrisCoe replied:
Good point. I had to replace `%d` with `%" PRIu32 "`.

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -64,33 +67,268 @@ void set_client_timer(struct itimerval *timer) {
+/*  Store frequencies for each non-empty bin. All bins include only start (exclusive) as
+    end (inclusive) can be inferred by adding bin size. Outlier bins include both
+    start and end of bin as size depends on outliers. */
+void storeHistogram(int binSize, std::map<int, int> &activeBins, int minValue, int maxValue) {
```
Add the `const` modifier to `activeBins` since map content is not altered here.

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -400,6 +400,17 @@ static int proc_mode_under_load(int id, int argc, const char **argv) {
+          AOPT_OPTARG,
+          aopt_set_literal(0),
+          aopt_set_string("histogram"),
+          "Build histogram of latencies. Optional histogram arguments formated as binsize:lowerrange:upperrange " },
```
Is the argument really optional? Missing optarg leads to parse error.

ChrisCoe replied:
This is outdated. Updated description and aopt type to reflect required args for histogram flag.

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -64,33 +67,268 @@ void set_client_timer(struct itimerval *timer) {
+    } else {
+        log_msg("[Histogram] Display scaled to fit on screen (Key: '#' = up to %d samples)", scalingUnit);
+    }
+    log_msg("%*s %*s", binsHeaderWidth, binsString.c_str(), frequencyHeaderWidth, freqString.c_str());
```
Two whitespaces before frequencies as in line 203:
```suggestion
    log_msg("%*s  %*s", binsHeaderWidth, binsString.c_str(), frequencyHeaderWidth, freqString.c_str());
```

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -64,33 +67,268 @@ void set_client_timer(struct itimerval *timer) {
+}
+
+//------------------------------------------------------------------------------
+/* Sparce fixed bin histogram with outlier bins outside given range */
```
Did you mean "Sparse"?

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -64,33 +67,268 @@ void set_client_timer(struct itimerval *timer) {
+        storeHistogram(binSize, activeBins, minValue, maxValue);
+    }
+
+    activeBins.clear();
```
Not needed since `activeBins` is a local variable.

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -126,16 +394,32 @@ void client_statistics(int serverNo, Message *pMsgRequest) {
+                            totalRunTime.toDecimalUsec() / 1000000,
+                            g_pApp->m_const_params.warmup_msec, sendCount, receiveCount, g_skipCount);
+            } else {
+            log_msg_file2(f, "[Total Run] RunTime=%.3lf sec; Warm up packets=%" PRIu64
```
```suggestion
                log_msg_file2(f, "[Total Run] RunTime=%.3lf sec; Warm up packets=%" PRIu64
```

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -441,11 +762,19 @@ void Client<IoType, SwitchDataIntegrity, SwitchActivityInfo, SwitchCycleDuration
+            } else {
             fprintf(f, "test was performed using the following parameters: "
-                       "--mps=%d --burst=%d --reply-every=%d --msg-size=%d --time=%d",
+                       "--mps=%d --burst=%d --reply-every=%d --msg-size=%d --number-of-packets=%" PRIu64 "",
```
`reply-every` parameter is not allowed in ping-pong operation mode where NUMBER_BASED measurement mode can be selected.

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -673,9 +1004,69 @@ template <class IoType, class SwitchDataIntegrity, class SwitchActivityInfo,
+                testStart = g_pPacketTimes->getTxTime(replyEvery); // first pong request packet
+            }
+
+            seqNo+= replyEvery;
```
Since in NUMBER_BASED mode replyEvery is always 1, this code may be simplified.

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -64,33 +67,268 @@ void set_client_timer(struct itimerval *timer) {
+    int terminalWidth = 0;
+    int scalingUnit = 0;
+    int maxDisplayWidth = 0;
+    int maxStartBinDigits = getNumberOfDigits(maxValue - binSize);
```
`maxValue` may be less or equal to `binSize`. In these cases `log10` returns NaN or -inf and number of digits is incorrect (< 0). This leads to "freeze" in column formatting code.
```suggestion
    int maxStartBinDigits = getNumberOfDigits(getStartOfRightOutlierBin());
```

ChrisCoe committed:
Address feedback

ChrisCoe committed:
Avoid negative number of digits and formatting

Co-authored-by: Alexander Galanin <91539962+agalanin-at-nvidia@users.noreply.github.com>

ChrisCoe committed:
Apply more suggestions from code review

Co-authored-by: Alexander Galanin <91539962+agalanin-at-nvidia@users.noreply.github.com>

ChrisCoe committed:
Merge with recent changes

agalanin-at-nvidia commented on src/aopt.cpp:
```diff
@@ -342,12 +340,12 @@ const AOPT_DESC *aopt_get_desc(const AOPT_DESC *aopt_desc, int key) {
     }
 
-    return *aopt_get_desc(aopt_desc, key)->longs;
+    return *aopt_desc->longs;
```
Result of `aopt_get_desc` is lost. You are always use the first aopt.

agalanin-at-nvidia commented on src/sockperf.cpp:
```diff
@@ -632,50 +632,20 @@ static int proc_mode_under_load(int id, int argc, const char **argv) {
+                /* Parse histogram options list */
+                long bin_size, lower_range, upper_range;
+                char suffix; //< needed to check for garbage at the end
+                if (sscanf(optarg, "%ld:%ld:%ld%c", &bin_size, &lower_range, &upper_range, &suffix) == required_args) {
```
Belated comment: you can use `PRIu32` also here and parse directly to `s_user_params.*` without using temporary variables.

ChrisCoe commented:
Hey, @agalanin-at-nvidia! I have addressed all your comments. Thank you, and let me know what you think 😊.

agalanin-at-nvidia commented on src/client.cpp:
```diff
@@ -762,19 +757,11 @@ void Client<IoType, SwitchDataIntegrity, SwitchActivityInfo, SwitchCycleDuration
                     (int)g_pApp->m_const_params.reply_every, (int)g_pApp->m_const_params.msg_size,
-                    g_pApp->m_const_params.number_test_target);
-            }
+                    (int)g_pApp->m_const_params.sec_test_duration);
```
NUMBER_BASED mode full-log output is broken after this change (-n 1000):
```
------------------------------
test was performed using the following parameters: --mps=600000 --burst=1 --reply-every=1 --msg-size=64 --time=1
------------------------------
```
Should display `--number-of-packets`, `--msg-size` and `--burst`.

ChrisCoe replied:
Got it. Made the change.

ChrisCoe committed:
Use aopt_get_desc retval and reduce temp vars

ChrisCoe committed:
Fix full-log output params display

agalanin-at-nvidia commented:
All my remarks have been fixed. @igor-ivanov, your turn!

igor-ivanov commented:
Moved to sockperf_v2 (https://github.com/Mellanox/sockperf/commit/a57ffb5790020021ea1dff2b6b7d9bf5b1669e6d)
